Winter: Skin tones with cooler undertones, often characterized by higher saturation and lower lightness values. Winter skin tones may have hues that range from cool blues to deep purples.
Spring: Skin tones with warm undertones and medium to high lightness values. Spring skin tones typically have a peachy or golden hue and moderate saturation.
Summer: Skin tones with cool undertones and moderate to high lightness values. Summer skin tones may have pink or rosy hues and varying levels of saturation.
Fall: Skin tones with warm undertones and moderate to low lightness values. Fall skin tones often have golden or bronze hues with moderate saturation.

Winter: Typically characterized by cooler undertones and higher saturation values. We'll look for lower lightness values and hues that may include cooler tones like blue, purple, or green.
Spring: Known for warm undertones and moderate saturation. We'll focus on lightness values that are not too high or low and hues that include warmer tones like peach, coral, or golden.
Summer: Features cool undertones and moderate to high lightness values. We'll look for hues that may include cooler tones like pink, lavender, or blue, and saturation that varies from moderate to high.
Fall: Identified by warm undertones and moderate saturation. We'll focus on moderate to low lightness values and hues that include warm tones like orange, brown, or red.

Winter:

Lightness: 0.4 - 0.7
Hue: 180 - 300
Saturation: 0.5 - 1.0
Spring:

Lightness: 0.5 - 0.8
Hue: 20 - 80
Saturation: 0.3 - 0.7
Summer:

Lightness: 0.6 - 0.9
Hue: 200 - 280
Saturation: 0.4 - 0.8
Fall:

Lightness: 0.3 - 0.6
Hue: 20 - 60
Saturation: 0.3 - 0.7

In [166]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Load the dataset
data = pd.read_csv('models.csv')


data

model         l     tone  n_covers
0               Adele  0.796199  #E3C1B2         2
1         Adwoa Aboah  0.486863  #BA673E         1
2     Alicia Vikander  0.756304  #E8B398         2
3        Amal Clooney  0.723736  #DBAF95         1
4     Amanda Seyfried  0.663791  #D59B7C         1
..                ...       ...      ...       ...
118       Uma Thurman  0.653213  #CEA57E         2
119    Vanessa Axente  0.674379  #DB997C         1
120  Vittoria Ceretti  0.621747  #DB8861         1
121      Winona Ryder  0.727591  #CEAFA4         1
122           Zendaya  0.598039  #C6876A         1

[123 rows x 4 columns]

the column labeled 'l' represents the lightness value of the corresponding hexadecimal color code ('tone' column). This value typically represents the perceived brightness of the color, where 0 corresponds to black and 1 corresponds to white in the HSL (Hue, Saturation, Lightness) color space.

In [167]:
# Load the dataset
data1 = pd.read_csv('faces.csv')


data1

date              model     tone         l
0     1/1/2018      Lupita Nyongo  #402C2C  0.212745
1    10/1/2015      Lupita Nyongo  #58342B  0.257423
2     3/1/2009     Michelle Obama  #844930  0.354118
3     4/1/2015    Serena Williams  #875038  0.375000
4     7/1/2014      Lupita Nyongo  #85524F  0.416667
..         ...                ...      ...       ...
257   9/1/2004  Natalia Vodianova  #F0D6C4  0.855556
258  12/1/2014          Amy Adams  #ECCDC9  0.857353
259  11/1/2018         Claire Foy  #EFD0C5  0.858227
260  12/1/2013   Jessica Chastain  #F8D2BE  0.861275
261  11/1/2010      Anne Hathaway  #F6D9CB  0.882353

[262 rows x 4 columns]

In [168]:
# Drop the "n_covers" column
data1.drop("date", axis=1, inplace=True)



In [169]:
data1

model     tone         l
0        Lupita Nyongo  #402C2C  0.212745
1        Lupita Nyongo  #58342B  0.257423
2       Michelle Obama  #844930  0.354118
3      Serena Williams  #875038  0.375000
4        Lupita Nyongo  #85524F  0.416667
..                 ...      ...       ...
257  Natalia Vodianova  #F0D6C4  0.855556
258          Amy Adams  #ECCDC9  0.857353
259         Claire Foy  #EFD0C5  0.858227
260   Jessica Chastain  #F8D2BE  0.861275
261      Anne Hathaway  #F6D9CB  0.882353

[262 rows x 3 columns]

In [170]:
#combine two dataset 

df = pd.concat([data, data1], axis= 0)

df

model         l     tone  n_covers
0                Adele  0.796199  #E3C1B2       2.0
1          Adwoa Aboah  0.486863  #BA673E       1.0
2      Alicia Vikander  0.756304  #E8B398       2.0
3         Amal Clooney  0.723736  #DBAF95       1.0
4      Amanda Seyfried  0.663791  #D59B7C       1.0
..                 ...       ...      ...       ...
257  Natalia Vodianova  0.855556  #F0D6C4       NaN
258          Amy Adams  0.857353  #ECCDC9       NaN
259         Claire Foy  0.858227  #EFD0C5       NaN
260   Jessica Chastain  0.861275  #F8D2BE       NaN
261      Anne Hathaway  0.882353  #F6D9CB       NaN

[385 rows x 4 columns]

In [171]:
df.drop("n_covers", axis=1, inplace=True)

In [172]:
df

model         l     tone
0                Adele  0.796199  #E3C1B2
1          Adwoa Aboah  0.486863  #BA673E
2      Alicia Vikander  0.756304  #E8B398
3         Amal Clooney  0.723736  #DBAF95
4      Amanda Seyfried  0.663791  #D59B7C
..                 ...       ...      ...
257  Natalia Vodianova  0.855556  #F0D6C4
258          Amy Adams  0.857353  #ECCDC9
259         Claire Foy  0.858227  #EFD0C5
260   Jessica Chastain  0.861275  #F8D2BE
261      Anne Hathaway  0.882353  #F6D9CB

[385 rows x 3 columns]

In [173]:
df.isnull().sum()

model    0
l        0
tone     0
dtype: int64

In [174]:
# Convert dataset to DataFrame
df1 = pd.DataFrame(df)

# Function to convert hex color code to RGB
def hex_to_rgb(tone):
    tone = tone.lstrip('#')
    return tuple(int(tone[i:i+2], 16) for i in (0, 2, 4))

# Apply function to convert hex color codes to RGB
df1['RGB'] = df1['tone'].apply(hex_to_rgb)

# Print DataFrame with RGB values
print(df1)

                 model         l     tone              RGB
0                Adele  0.796199  #E3C1B2  (227, 193, 178)
1          Adwoa Aboah  0.486863  #BA673E   (186, 103, 62)
2      Alicia Vikander  0.756304  #E8B398  (232, 179, 152)
3         Amal Clooney  0.723736  #DBAF95  (219, 175, 149)
4      Amanda Seyfried  0.663791  #D59B7C  (213, 155, 124)
..                 ...       ...      ...              ...
257  Natalia Vodianova  0.855556  #F0D6C4  (240, 214, 196)
258          Amy Adams  0.857353  #ECCDC9  (236, 205, 201)
259         Claire Foy  0.858227  #EFD0C5  (239, 208, 197)
260   Jessica Chastain  0.861275  #F8D2BE  (248, 210, 190)
261      Anne Hathaway  0.882353  #F6D9CB  (246, 217, 203)

[385 rows x 4 columns]


In [175]:
# rename to make it clearer for viewers 

df1.rename(columns={'tone': 'hexcode'}, inplace=True)
df1.rename(columns={'l': 'lightness'}, inplace=True)

In [176]:
df1

model  lightness  hexcode              RGB
0                Adele   0.796199  #E3C1B2  (227, 193, 178)
1          Adwoa Aboah   0.486863  #BA673E   (186, 103, 62)
2      Alicia Vikander   0.756304  #E8B398  (232, 179, 152)
3         Amal Clooney   0.723736  #DBAF95  (219, 175, 149)
4      Amanda Seyfried   0.663791  #D59B7C  (213, 155, 124)
..                 ...        ...      ...              ...
257  Natalia Vodianova   0.855556  #F0D6C4  (240, 214, 196)
258          Amy Adams   0.857353  #ECCDC9  (236, 205, 201)
259         Claire Foy   0.858227  #EFD0C5  (239, 208, 197)
260   Jessica Chastain   0.861275  #F8D2BE  (248, 210, 190)
261      Anne Hathaway   0.882353  #F6D9CB  (246, 217, 203)

[385 rows x 4 columns]

In [177]:
import matplotlib.colors as mcolors

df2 = pd.DataFrame(df1)

# Function to convert RGB to hue
def rgb_to_hue(rgb):
    hsv_tuple = mcolors.rgb_to_hsv([x/255.0 for x in rgb])
    return hsv_tuple[0] * 360

# Apply the function to each row and add a new column
df2['hue'] = df2['RGB'].apply(rgb_to_hue)

print(df2)

                 model  lightness  hexcode              RGB        hue
0                Adele   0.796199  #E3C1B2  (227, 193, 178)  18.367347
1          Adwoa Aboah   0.486863  #BA673E   (186, 103, 62)  19.838710
2      Alicia Vikander   0.756304  #E8B398  (232, 179, 152)  20.250000
3         Amal Clooney   0.723736  #DBAF95  (219, 175, 149)  22.285714
4      Amanda Seyfried   0.663791  #D59B7C  (213, 155, 124)  20.898876
..                 ...        ...      ...              ...        ...
257  Natalia Vodianova   0.855556  #F0D6C4  (240, 214, 196)  24.545455
258          Amy Adams   0.857353  #ECCDC9  (236, 205, 201)   6.857143
259         Claire Foy   0.858227  #EFD0C5  (239, 208, 197)  15.714286
260   Jessica Chastain   0.861275  #F8D2BE  (248, 210, 190)  20.689655
261      Anne Hathaway   0.882353  #F6D9CB  (246, 217, 203)  19.534884

[385 rows x 5 columns]


In [178]:


# Define the hex_to_saturation function
def hex_to_saturation(hexcode):
    # Convert hexcode to RGB decimal values
    r = int(hexcode[1:3], 16) / 255
    g = int(hexcode[3:5], 16) / 255
    b = int(hexcode[5:7], 16) / 255

    # Find max and min of RGB values
    max_val = max(r, g, b)
    min_val = min(r, g, b)

    # Calculate saturation
    if max_val != min_val:
        saturation = (max_val - min_val) / (1 - abs(1 - (max_val + min_val)))
    else:
        saturation = 0

    return saturation



# Add a new column 'saturation' based on the 'hexcode' column
df2['saturation'] = df2['hexcode'].apply(hex_to_saturation)

# Display the updated dataset with the new 'saturation' column
print(df2)


                 model  lightness  hexcode              RGB        hue  \
0                Adele   0.796199  #E3C1B2  (227, 193, 178)  18.367347   
1          Adwoa Aboah   0.486863  #BA673E   (186, 103, 62)  19.838710   
2      Alicia Vikander   0.756304  #E8B398  (232, 179, 152)  20.250000   
3         Amal Clooney   0.723736  #DBAF95  (219, 175, 149)  22.285714   
4      Amanda Seyfried   0.663791  #D59B7C  (213, 155, 124)  20.898876   
..                 ...        ...      ...              ...        ...   
257  Natalia Vodianova   0.855556  #F0D6C4  (240, 214, 196)  24.545455   
258          Amy Adams   0.857353  #ECCDC9  (236, 205, 201)   6.857143   
259         Claire Foy   0.858227  #EFD0C5  (239, 208, 197)  15.714286   
260   Jessica Chastain   0.861275  #F8D2BE  (248, 210, 190)  20.689655   
261      Anne Hathaway   0.882353  #F6D9CB  (246, 217, 203)  19.534884   

     saturation  
0      0.466667  
1      0.500000  
2      0.634921  
3      0.492958  
4      0.514451  
.. 

In [179]:
df2.nunique() #

model         123
lightness     321
hexcode       323
RGB           323
hue           301
saturation    312
dtype: int64

In [180]:
df2

model  lightness  hexcode              RGB        hue  \
0                Adele   0.796199  #E3C1B2  (227, 193, 178)  18.367347   
1          Adwoa Aboah   0.486863  #BA673E   (186, 103, 62)  19.838710   
2      Alicia Vikander   0.756304  #E8B398  (232, 179, 152)  20.250000   
3         Amal Clooney   0.723736  #DBAF95  (219, 175, 149)  22.285714   
4      Amanda Seyfried   0.663791  #D59B7C  (213, 155, 124)  20.898876   
..                 ...        ...      ...              ...        ...   
257  Natalia Vodianova   0.855556  #F0D6C4  (240, 214, 196)  24.545455   
258          Amy Adams   0.857353  #ECCDC9  (236, 205, 201)   6.857143   
259         Claire Foy   0.858227  #EFD0C5  (239, 208, 197)  15.714286   
260   Jessica Chastain   0.861275  #F8D2BE  (248, 210, 190)  20.689655   
261      Anne Hathaway   0.882353  #F6D9CB  (246, 217, 203)  19.534884   

     saturation  
0      0.466667  
1      0.500000  
2      0.634921  
3      0.492958  
4      0.514451  
..          ...  
257    0.594595  
258    0.479452  
259    0.567568  
260    0.805556  
261    0.704918  

[385 rows x 6 columns]

In [181]:
##25 OF MARCH ATTEMPT 
# Define thresholds for each season based on common knowledge about seasonal color analysis
spring_ranges = {
    'hue': (0, 90),         # Warm hues
    'saturation': (0.3, 1),  # Moderate to high saturation
    'lightness': (0.4, 0.7)  # Moderate lightness
}

summer_ranges = {
    'hue': (90, 180),       # Cool hues
    'saturation': (0.3, 1),  # Moderate to high saturation
    'lightness': (0.5, 0.8)  # Moderate lightness
}

autumn_ranges = {
    'hue': (0, 90),         # Warm hues
    'saturation': (0.3, 1),  # Moderate to high saturation
    'lightness': (0.7, 1)    # High lightness
}

winter_ranges = {
    'hue': (180, 360),      # Cool hues
    'saturation': (0.3, 1),  # Moderate to high saturation
    'lightness': (0, 0.6)    # Low to moderate lightness
}

# Function to classify individuals into seasonal color palettes
def classify_season(row):
    hue, saturation, lightness = row['hue'], row['saturation'], row['lightness']
    
    if spring_ranges['hue'][0] <= hue <= spring_ranges['hue'][1] \
        and spring_ranges['saturation'][0] <= saturation <= spring_ranges['saturation'][1] \
        and spring_ranges['lightness'][0] <= lightness <= spring_ranges['lightness'][1]:
        return 'Spring'
    
    elif summer_ranges['hue'][0] <= hue <= summer_ranges['hue'][1] \
        and summer_ranges['saturation'][0] <= saturation <= summer_ranges['saturation'][1] \
        and summer_ranges['lightness'][0] <= lightness <= summer_ranges['lightness'][1]:
        return 'Summer'
    
    elif autumn_ranges['hue'][0] <= hue <= autumn_ranges['hue'][1] \
        and autumn_ranges['saturation'][0] <= saturation <= autumn_ranges['saturation'][1] \
        and autumn_ranges['lightness'][0] <= lightness <= autumn_ranges['lightness'][1]:
        return 'Autumn'
    
    elif winter_ranges['hue'][0] <= hue <= winter_ranges['hue'][1] \
        and winter_ranges['saturation'][0] <= saturation <= winter_ranges['saturation'][1] \
        and winter_ranges['lightness'][0] <= lightness <= winter_ranges['lightness'][1]:
        return 'Winter'
    
    else:
        return 'Unknown'

# Apply classification function to each row
df2['season'] = df2.apply(classify_season, axis=1)

# Print the counts of individuals in each seasonal color palette
print(df2['season'].value_counts())


season
Spring     215
Autumn     128
Unknown     42
Name: count, dtype: int64


def calc_dis(rgb_codes):
    spring = [[253,183,169],[247,98,77],[186,33,33]]
    summer = [[243,184,202],[211,118,155],[147,70,105]]
    autum = [[210,124,110],[155,70,60],[97,16,28]]
    winter = [[237,223,227],[177,47,57],[98,14,37]]
    

In [182]:
df1

model  lightness  hexcode              RGB
0                Adele   0.796199  #E3C1B2  (227, 193, 178)
1          Adwoa Aboah   0.486863  #BA673E   (186, 103, 62)
2      Alicia Vikander   0.756304  #E8B398  (232, 179, 152)
3         Amal Clooney   0.723736  #DBAF95  (219, 175, 149)
4      Amanda Seyfried   0.663791  #D59B7C  (213, 155, 124)
..                 ...        ...      ...              ...
257  Natalia Vodianova   0.855556  #F0D6C4  (240, 214, 196)
258          Amy Adams   0.857353  #ECCDC9  (236, 205, 201)
259         Claire Foy   0.858227  #EFD0C5  (239, 208, 197)
260   Jessica Chastain   0.861275  #F8D2BE  (248, 210, 190)
261      Anne Hathaway   0.882353  #F6D9CB  (246, 217, 203)

[385 rows x 4 columns]

In [183]:
#RBG ATTEMPT


# Extract RGB components from the 'RGB' column
df1['red'] = df1['RGB'].apply(lambda x: x[0])  # Extract red component
df1['green'] = df1['RGB'].apply(lambda x: x[1])  # Extract green component
df1['blue'] = df1['RGB'].apply(lambda x: x[2])  # Extract blue component

# Define threshold ranges for each season based on RGB values
spring_ranges_rgb = {
    'red': (210, 250),     # High red
    'green': (150, 217),   # Moderate green
    'blue': (148, 205)     # Moderate blue
}

summer_ranges_rgb = {
    'red': (100, 200),   # Moderate red
    'green': (0, 250),   # High green
    'blue': (0, 250)     # High blue
}

autumn_ranges_rgb = {
    'red': (210, 255),     # High red
    'green': (120, 190),   # Moderate green
    'blue': (0, 150)       # Low to moderate blue
}

winter_ranges_rgb = {
    'red': (0, 250),     # Low to moderate red
    'green': (0, 180),   # Moderate green
    'blue': (0, 255)     # High blue
}

# Function to classify individuals into seasonal color palettes based on RGB values
def classify_season_rgb(row):
    red, green, blue = row['red'], row['green'], row['blue']
    
    if spring_ranges_rgb['red'][0] <= red <= spring_ranges_rgb['red'][1] \
        and spring_ranges_rgb['green'][0] <= green <= spring_ranges_rgb['green'][1] \
        and spring_ranges_rgb['blue'][0] <= blue <= spring_ranges_rgb['blue'][1]:
        return 'Spring'
    
    elif summer_ranges_rgb['red'][0] <= red <= summer_ranges_rgb['red'][1] \
        and summer_ranges_rgb['green'][0] <= green <= summer_ranges_rgb['green'][1] \
        and summer_ranges_rgb['blue'][0] <= blue <= summer_ranges_rgb['blue'][1]:
        return 'Summer'
    
    elif autumn_ranges_rgb['red'][0] <= red <= autumn_ranges_rgb['red'][1] \
        and autumn_ranges_rgb['green'][0] <= green <= autumn_ranges_rgb['green'][1] \
        and autumn_ranges_rgb['blue'][0] <= blue <= autumn_ranges_rgb['blue'][1]:
        return 'Autumn'
    
    elif winter_ranges_rgb['red'][0] <= red <= winter_ranges_rgb['red'][1] \
        and winter_ranges_rgb['green'][0] <= green <= winter_ranges_rgb['green'][1] \
        and winter_ranges_rgb['blue'][0] <= blue <= winter_ranges_rgb['blue'][1]:
        return 'Winter'
    
    else:
        return 'Unknown'

# Apply classification function to each row
df1['season_rgb'] = df1.apply(classify_season_rgb, axis=1)

# Print the counts of individuals in each seasonal color palette based on RGB values
print(df1['season_rgb'].value_counts())


season_rgb
Summer     125
Spring     100
Autumn      91
Winter      68
Unknown      1
Name: count, dtype: int64


In [184]:
unknown_rows = df1[df1['season_rgb'] == 'Unknown']

# Print the rows where 'unknown' values exist
print("Rows with 'unknown' values:")
print(unknown_rows)

Rows with 'unknown' values:
           model  lightness  hexcode              RGB  red  green  blue  \
214  Rooney Mara   0.746588  #D1B6AB  (209, 182, 171)  209    182   171   

    season_rgb  
214    Unknown  


Rooney Mara (RGB: 209, 182, 171):

This color has moderate red (209), moderate green (182), and moderate blue (171) components. It's a muted, desaturated tone with balanced RGB values, making it difficult to pinpoint a specific seasonal association.

In [185]:
# Given classification criteria for each season
season_criteria = {
    'Winter': {'lightness': (0.2, 0.8), 'hue': (140, 320), 'saturation': (0.3, 1.0)},
    'Spring': {'lightness': (0.3, 0.9), 'hue': (0, 120), 'saturation': (0.1, 0.9)},
    'Summer': {'lightness': (0.4, 1.0), 'hue': (140, 300), 'saturation': (0.2, 1.0)},
    'Autumn': {'lightness': (0.1, 0.7), 'hue': (0, 100), 'saturation': (0.1, 0.8)}}

# Function to classify a color based on given criteria
def classify_season(lightness, hue, saturation):
    for season, criteria in season_criteria.items():
        lightness_range = criteria['lightness']
        hue_range = criteria['hue']
        saturation_range = criteria['saturation']
        if lightness_range[0] <= lightness <= lightness_range[1] and \
           hue_range[0] <= hue <= hue_range[1] and \
           saturation_range[0] <= saturation <= saturation_range[1]:
            return season
    return 'Unknown'

# Apply classification to each entry in the dataset
df2['season'] = df2.apply(lambda row: classify_season(row['lightness'], row['hue'], row['saturation']), axis=1)

# Display the seasonal classification for all entries
print(df2[['model', 'season']])


                 model  season
0                Adele  Spring
1          Adwoa Aboah  Spring
2      Alicia Vikander  Spring
3         Amal Clooney  Spring
4      Amanda Seyfried  Spring
..                 ...     ...
257  Natalia Vodianova  Spring
258          Amy Adams  Spring
259         Claire Foy  Spring
260   Jessica Chastain  Spring
261      Anne Hathaway  Spring

[385 rows x 2 columns]


In [186]:
df2['season'].value_counts()


season
Spring    383
Autumn      2
Name: count, dtype: int64

In [187]:
df1

model  lightness  hexcode              RGB  red  green  blue  \
0                Adele   0.796199  #E3C1B2  (227, 193, 178)  227    193   178   
1          Adwoa Aboah   0.486863  #BA673E   (186, 103, 62)  186    103    62   
2      Alicia Vikander   0.756304  #E8B398  (232, 179, 152)  232    179   152   
3         Amal Clooney   0.723736  #DBAF95  (219, 175, 149)  219    175   149   
4      Amanda Seyfried   0.663791  #D59B7C  (213, 155, 124)  213    155   124   
..                 ...        ...      ...              ...  ...    ...   ...   
257  Natalia Vodianova   0.855556  #F0D6C4  (240, 214, 196)  240    214   196   
258          Amy Adams   0.857353  #ECCDC9  (236, 205, 201)  236    205   201   
259         Claire Foy   0.858227  #EFD0C5  (239, 208, 197)  239    208   197   
260   Jessica Chastain   0.861275  #F8D2BE  (248, 210, 190)  248    210   190   
261      Anne Hathaway   0.882353  #F6D9CB  (246, 217, 203)  246    217   203   

    season_rgb  
0       Spring  
1       Summer  
2       Spring  
3       Spring  
4       Autumn  
..         ...  
257     Spring  
258     Spring  
259     Spring  
260     Spring  
261     Spring  

[385 rows x 8 columns]

In [188]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score, roc_curve, auc


In [189]:
# Feature engineering
# You might need to convert hex codes to RGB values or perform other transformations


#LOGISTIC REGRESSION 


# Split the data into features and target variable
X = df1[['red', 'green', 'blue']]
y = df1['season_rgb'] #target

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize and train the model
model = LogisticRegression()
model.fit(X_train, y_train)

# Make predictions on the testing set
y_pred = model.predict(X_test)

# Calculate evaluation metrics
accuracy = accuracy_score(y_test, y_pred)

print("Logistic Regression Accuracy:", accuracy)

Logistic Regression Accuracy: 0.4935064935064935


/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [190]:

#Example of predicting the seasonal color palette for a new skin tone sample
new_sample = [[20, 50, 200]]  # RGB values for the new skin tone 
#in this case i want to add user image 
predicted_palette = model.predict(new_sample)
print("Predicted seasonal color palette:", predicted_palette)


Predicted seasonal color palette: ['Unknown']


/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


In [191]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# Assume 'X' contains HSL values and 'y' contains seasonal color palette labels
# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=385)

# Initialize the RandomForestClassifier
clf = RandomForestClassifier(n_estimators=100, random_state=385)     # change this around 

# Train the classifier
clf.fit(X_train, y_train)

# Make predictions on the test set
y_pred = clf.predict(X_test)

# Evaluate the classifier
accuracy = accuracy_score(y_test, y_pred)
print("Random Forest Accuracy:", accuracy)

#overfitting 


Random Forest Accuracy: 0.9827586206896551


In [192]:
#Example of predicting the seasonal color palette for a new skin tone sample
new_sample = [[227, 193, 178]]  # RGB values for the new skin tone 
#in this case i want to add user image 
predicted_palette = clf.predict(new_sample) #clf random forest
print("Predicted seasonal color palette:", predicted_palette)


Predicted seasonal color palette: ['Spring']


/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


In [193]:
df1

model  lightness  hexcode              RGB  red  green  blue  \
0                Adele   0.796199  #E3C1B2  (227, 193, 178)  227    193   178   
1          Adwoa Aboah   0.486863  #BA673E   (186, 103, 62)  186    103    62   
2      Alicia Vikander   0.756304  #E8B398  (232, 179, 152)  232    179   152   
3         Amal Clooney   0.723736  #DBAF95  (219, 175, 149)  219    175   149   
4      Amanda Seyfried   0.663791  #D59B7C  (213, 155, 124)  213    155   124   
..                 ...        ...      ...              ...  ...    ...   ...   
257  Natalia Vodianova   0.855556  #F0D6C4  (240, 214, 196)  240    214   196   
258          Amy Adams   0.857353  #ECCDC9  (236, 205, 201)  236    205   201   
259         Claire Foy   0.858227  #EFD0C5  (239, 208, 197)  239    208   197   
260   Jessica Chastain   0.861275  #F8D2BE  (248, 210, 190)  248    210   190   
261      Anne Hathaway   0.882353  #F6D9CB  (246, 217, 203)  246    217   203   

    season_rgb  
0       Spring  
1       Summer  
2       Spring  
3       Spring  
4       Autumn  
..         ...  
257     Spring  
258     Spring  
259     Spring  
260     Spring  
261     Spring  

[385 rows x 8 columns]

In [194]:

from sklearn.naive_bayes import GaussianNB


In [195]:
 ## naives bayes
model_NB = GaussianNB()
model_NB.fit(X_train, y_train)

GaussianNB()

In [196]:
## predict the test set
y_pred_NB = model_NB.predict(X_test)
pd.DataFrame({'Actual': y_test, 'Predicted': y_pred_NB})

Actual Predicted
180  Spring    Autumn
18   Spring    Autumn
79   Summer    Summer
103  Summer    Winter
167  Autumn    Spring
..      ...       ...
101  Autumn    Autumn
32   Summer    Summer
48   Autumn    Autumn
4    Autumn    Autumn
188  Autumn    Spring

[116 rows x 2 columns]

In [197]:
## Evaluate the naive bayes model
accuracy_NB = accuracy_score(y_test, y_pred_NB)

print("NB Accuracy Score:", accuracy_NB)

NB Accuracy Score: 0.8448275862068966


In [202]:
#Example of predicting the seasonal color palette for a new skin tone sample
new_sample = [[213, 155, 124]]  # RGB values for the new skin tone 
#in this case i want to add user image 
predicted_palette = model_NB.predict(new_sample)
print("Predicted seasonal color palette:", predicted_palette)

Predicted seasonal color palette: ['Autumn']


/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


In [199]:
df1

model  lightness  hexcode              RGB  red  green  blue  \
0                Adele   0.796199  #E3C1B2  (227, 193, 178)  227    193   178   
1          Adwoa Aboah   0.486863  #BA673E   (186, 103, 62)  186    103    62   
2      Alicia Vikander   0.756304  #E8B398  (232, 179, 152)  232    179   152   
3         Amal Clooney   0.723736  #DBAF95  (219, 175, 149)  219    175   149   
4      Amanda Seyfried   0.663791  #D59B7C  (213, 155, 124)  213    155   124   
..                 ...        ...      ...              ...  ...    ...   ...   
257  Natalia Vodianova   0.855556  #F0D6C4  (240, 214, 196)  240    214   196   
258          Amy Adams   0.857353  #ECCDC9  (236, 205, 201)  236    205   201   
259         Claire Foy   0.858227  #EFD0C5  (239, 208, 197)  239    208   197   
260   Jessica Chastain   0.861275  #F8D2BE  (248, 210, 190)  248    210   190   
261      Anne Hathaway   0.882353  #F6D9CB  (246, 217, 203)  246    217   203   

    season_rgb  
0       Spring  
1       Summer  
2       Spring  
3       Spring  
4       Autumn  
..         ...  
257     Spring  
258     Spring  
259     Spring  
260     Spring  
261     Spring  

[385 rows x 8 columns]